### ISSUE TITLE
[SPIR-V] template specialization

### ISSUE THREAD

**Issue by Przemog1** [microsoft/DirectXShaderCompiler#5859]
**Description**
[Assertion](https://github.com/microsoft/DirectXShaderCompiler/blob/main/tools/clang/lib/SPIRV/DeclResultIdMapper.cpp#L1029) fails when I'm trying to compile code containing template specialization.

**Steps to Reproduce**
https://godbolt.org/z/cbjv3ffGx
```cpp
template<uint16_t bits_log2>
struct clz_masks
{
    static const uint16_t SHIFT = uint16_t(1)<<(bits_log2-1);
    static const uint64_t LO_MASK = (1ull<<SHIFT)-1;
};

// all good if i remove code below:
template<>
struct clz_masks<0>
{
    static const uint16_t shift = 0;
    static const uint64_t lo_mask = 0;
};
```

**Actual Behavior**
[Assertion](https://github.com/microsoft/DirectXShaderCompiler/blob/main/tools/clang/lib/SPIRV/DeclResultIdMapper.cpp#L1029) fails.

**Environment**
- DXC version libdxcompiler.so: 1.7(dev;1-d6e7e37e)
- Host Operating System Windows but DXC as a lib

**Comment by devshgraphicsprogramming** [comment]
this could be caused by the same thing as #5826 except that the compiler runs into an assert not triggered in Release mode, so godbolt doesn't crash.

**Comment by devshgraphicsprogramming** [comment]
btw this super basic example also tridders the assert
```cpp
template<bool val>
struct bool_type
{
    static const bool value = true;
};

// all good if i remove code below:
template<>
struct bool_type<false>
{
    static const bool value = false;
};
```

looks like there's a collision and the `var` has the same raw-pointer value for both `bool_type::value` and `bool_type<false>::value`
https://github.com/microsoft/DirectXShaderCompiler/blob/93ad5b31316cfed379a8e83bb6cfbbc8bc886ff9/tools/clang/lib/SPIRV/DeclResultIdMapper.cpp#L1029

**Comment by devshgraphicsprogramming** [comment]
Mentioning original author @ehsannas

This might also underly #5823 and #5821

P.S. @cassiebeckley maybe you want to look into this?
